### PR TITLE
Enh/minimal projection system

### PR DESCRIPTION
**Pull request by kalmarek**
oh boy, it turned out larger than i expected.

The main addition is the `semisimple=false` flag to `symmetry_adapted_basis` which tries "best effort" procedure to determine the minimal projection system = isomorphism to the simple part of the appropriate group *-algebra.
The difference can be observed in `https://github.com/kalmarek/SymbolicWedderburn.jl/blob/enh/minimal_projection_system/test/action_permutation.jl`: with `semisimple=true` it gives a `[80, 22, 18]`-block decomposition with degrees `[2,1,1]`. With `semisimple=false` a `[40, 22, 18]`-decomposition is returned. This procedure works very fast for my examples (`C₂≀S₅` of order `3840`), but doesn't necessarily work for all the groups in `SmallPermGroups`. You'll see lots of warnings during the tests.

Notable changes:
* `Character`s are based on `CharacterTable`s now.
* all computations is done with `Rational{<:Integer}` up to constructing the matrix projections, where the type `T` provided by the user determines the `eltype` of matrices. E.g. if you know that all characters of your group (that occur in the action on a given basis) are `Rational`-valued you may ask for `Rational{Int}`-valued `symmetry_adapted_basis`
* I added `CachedExtensionHomomorphism` which caches its values in a dict. for larger groups and bases this makes huge difference. Also this solves the `inv(Matrix(g))` hack in one place

lots of other small things were implemented/enhanced on the way:

TODO:
 - [x] find the `rank_one_projection` for realified characters `χ + conj(χ)`.
 - [x] remove warnings?
 - [x] Separate `Character`s/`CharacterTable`s to a submodule.
 - [x] rework file-structure.
 - [x] Add Robinson form as test case to test `ByLinearTransformation` 

@blegat

### CODE REVIEW

File: src/sa_basis.jl [hunk 1]
```diff
@@ -1,160 +1,233 @@
...
-Compute a basis for the linear space `ℝⁿ` which is invariant under the symmetry of `G`.
+    symmetry_adapted_basis([T::Type,] G::AbstractPermutationGroup[, S=Rational{Int};
+        semisimple=false])
+Compute a basis for the linear space `ℝⁿ` (where `n = degree(G)`) which is
+invariant under the symmetry of `G`.
 
-The permutation group is acting naturally on `1:degree(G)`. The coefficients of
-the invariant basis are returned in (orthogonal) blocks corresponding to irreducible
-characters of `G`.
+The coefficients of the invariant basis are returned in (orthogonal) blocks
```
Comment (blegat) [inline src/sa_basis.jl:73]: Aren't they orthogonal only if `T<:AbstractFloat` ?

File: src/sa_basis.jl [hunk 2]
```diff
@@ -1,160 +1,233 @@
...
+summands. Otherwise semisimple decomposition is computed.
 
 !!! Note:
-Each block is invariant under the action of `G`, i.e. the action may permute
-vectors from symmetry adapted basis within each block. The blocks are guaranteed
-to be orthogonal. If `T<:LinearAlgebra.BlasFloat` BLAS routines will be used to
-orthogonalize vectors within each block.
+Each returned block (a `DirectSummand`) is invariant under the action of `G`,
+which means that the action may still e.g. permute (row) vectors , but only
+*within* each block. The blocks are guaranteed to be orthogonal.
```
Comment (blegat) [inline src/sa_basis.jl:88]: Same here

File: src/sa_basis.jl (hunk 1 above)
Comment (kalmarek) [reply src/sa_basis.jl:73]: The blocks are always pairwise orthogonal. The vectors within each block are orthogonal only when `T<:AbstractFloat`

File: src/actions.jl [hunk 3]
```diff
@@ -2,132 +2,123 @@ abstract type Action end
...
 
+induce(hom::InducedActionHomomorphism, g::GroupElement) = induce(action(hom), hom, g)
+
 decompose(x::Any, hom::InducedActionHomomorphism) = throw(
     "No fallback is provided for $(typeof(x)). You need to implement `decompose(::$(typeof(x)), ::$(typeof(hom)))`.",
 )
 
-coeff_type(ac::ByLinearTransformation) = throw(
-    "No fallback is provided for $(typeof(ac)). You need to implement `coeff_type(::$(typeof(ac)))`.",
+_coeff_type(ac::ByLinearTransformation) = throw(
```
Comment (blegat) [inline src/actions.jl:25]: This is part of the interface and is implemented in SumOfSquares examples so it should not start with an `_`

File: src/sa_basis.jl (hunk 1 above)
Comment (blegat) [reply src/sa_basis.jl:73]: Maybe that could be clearer in the docstring

**Comment by blegat** [comment]
I posted failures in https://github.com/jump-dev/SumOfSquares.jl/pull/207

**Comment by kalmarek** [comment]
ok, i'll look into these

File: src/actions.jl (hunk 3 above)
Comment (kalmarek) [reply src/actions.jl:25]: yeah, I forgot this one is part of the interface ;/ I'm not very happy about this, but I'm going to reverse the change ;)

**Comment by kalmarek** [comment]
@blegat I force-pushed so if you cloned this, you'll need to force pull;

This seems to be ready (that's correct: READY!) on my side ;)
Though probably should have been 3 or 4 PRs...

but anyway: I shuffled some of the names and exported others;
Notable name changes that you might be interfacing with are:
 * `SymbolicWedderburn.basis(ds::DirectSummand)` → `SymbolicWedderburn.image_basis(ds::DirectSummand)` (`basis` has very different meaning in the package)
 * I decided to export `symmetry_adapted_basis`, `WedderburnDecomposition`, `basis`, `direct_summands`, `invariant_vectors`, `issimple`, and `multiplicity`. I don't export `degree` (from `PermutationGroups`, since that's too generic name). Let me know if you want something more exported.

----

The general use-case pattern should is as follows:

```julia
C = ... # objective vector
G = ... # a group
action = ... # instance of action::SymbolicWedderburn.Action
T = Float64 # mandatory at the moment

basis_full = ... # basis in which C is written, and G acts on
basis_half = ... # basis for psd constraint, e.g. monomials(x, 0:(DynamicPolynomials.maxdegree(f) ÷ 2))


wedderburn = WedderburnDecomposition(T, G, action, basis_full, basis_half)

let m = JuMP.Model()
    psds = [
        JuMP.@variable(m, [1:d, 1:d] in PSDCone())
        for d in size.(direct_summands(wedderburn), 1)
    ]

    # preallocating
    Mπs = zeros.(T, size.(psds))
    M_orb = let d = size(first(direct_summands(wedderburn)), 2)
        zeros(T, d, d)
    end
    tmps = zeros.(T, reverse.(size.(direct_summands(wedderburn))))

    for iv in invariant_vectors(wedderburn)
        c = dot(C, iv)
        M_orb = invariant_constraint!(M_orb, M, iv)
        Mπs = SymbolicWedderburn.diagonalize!(Mπs, M_orb, wedderburn, tmps)
        JuMP.@constraint m sum(
            dot(Mπ, Pπ) for (Mπ, Pπ) in zip(Mπs, psds) if !iszero(Mπ)
        ) == c
    end
    m
end

optimize!(m)
```
here `invariant_constraint!(M_orb, M, iv)` is defined locally; it essentially amounts to averaging constraint matrices `M` (indexed by the elements of `basis_full`) with the weights provided by `iv` and storing the result in (dense?) matrix `M_orb`. see `examples/util.jl` for an example where `M[i,j] = k` whenever `basis_half[i]*basis_half[j] == basis_full[k]`. But I think you need to ship your own ;)

**Comment by kalmarek** [comment]
@blegat  I finally understood what minimal rank projection for `χ + conj(χ)` should look like ;)
(doing anything but recording the talk :P)

there are about ~ 20 characters in the small groups when we can't find the minimal rank projection;
I have an idea how to make it better, but this would need much code and in general is kind of open research problem ;)

----
```
[ Info: SmallGroup(8, 4)
┌ Warning: Could not find minimal projection for χ₁
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
[ Info: SmallGroup(12, 1)
┌ Warning: Could not find minimal projection for χ₂
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
[ Info: SmallGroup(16, 4)
┌ Warning: Could not find minimal projection for χ₂
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
[ Info: SmallGroup(16, 9)
┌ Warning: Could not find minimal projection for χ₂
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
┌ Warning: Could not find minimal projection for χ₁
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
[ Info: SmallGroup(16, 12)
┌ Warning: Could not find minimal projection for χ₂
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
┌ Warning: Could not find minimal projection for χ₁
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
[ Info: SmallGroup(20, 1)
┌ Warning: Could not find minimal projection for χ₄
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
┌ Warning: Could not find minimal projection for χ₃
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
[ Info: SmallGroup(24, 1)
┌ Warning: Could not find minimal projection for χ₂ +χ₃
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
┌ Warning: Could not find minimal projection for χ₄
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
[ Info: SmallGroup(24, 3)
┌ Warning: Could not find minimal projection for χ₆
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
[ Info: SmallGroup(24, 4)
┌ Warning: Could not find minimal projection for χ₁
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
┌ Warning: Could not find minimal projection for χ₂
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
┌ Warning: Could not find minimal projection for χ₃
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
[ Info: SmallGroup(24, 7)
┌ Warning: Could not find minimal projection for χ₂
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
┌ Warning: Could not find minimal projection for χ₄
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
[ Info: SmallGroup(24, 11)
┌ Warning: Could not find minimal projection for χ₁
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
┌ Warning: Could not find minimal projection for χ₂ +χ₃
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
[ Info: SmallGroup(28, 1)
┌ Warning: Could not find minimal projection for χ₅
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
┌ Warning: Could not find minimal projection for χ₄
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122
┌ Warning: Could not find minimal projection for χ₆
└ @ SymbolicWedderburn ~/.julia/dev/SymbolicWedderburn/src/minimal_projections.jl:122

```

**Pull request merged**